tests/30rooms/13anonymousaccess.pl

```diff
@@ -0,0 +1,92 @@
+test "Anonymous user cannot view non-world-readable rooms",
+   requires => [ qw( first_api_client ), local_user_preparer() ],
+
+   do => sub {
+      my ( $api_client, $user ) = @_;
+
+      my $anonymous_user;
+      my $room_id;
+
+      register_anonymous_user( $api_client )->then( sub {
+         ( $anonymous_user ) = @_;
+
+         matrix_create_and_join_room( [ $user ] )
+         ->then( sub {
+            ( $room_id ) = @_;
+
+            set_room_history_visibility( $user, $room_id, "shared" );
+         })->then( sub {
+            matrix_send_room_text_message( $user, $room_id, body => "mice" )
+         })->then( sub {
+            do_request_json_for( $anonymous_user,
+               method => "GET",
+               uri    => "/api/v1/rooms/$room_id/messages",
+               params => {
+                  limit => "1",
+                  dir   => "b",
+               },
+            )
+         })->main::expect_http_403;
+      });
+   };
+
+test "Anonymous user can view world-readable rooms",
+   requires => [ qw( first_api_client ), local_user_preparer() ],
+
+   do => sub {
+      my ( $api_client, $user ) = @_;
+
+      my $anonymous_user;
+      my $room_id;
+
+      register_anonymous_user( $api_client )->then( sub {
+         ( $anonymous_user ) = @_;
+
+         matrix_create_and_join_room( [ $user ] )
+         ->then( sub {
+            ( $room_id ) = @_;
+
+            set_room_history_visibility( $user, $room_id, "world_readable" );
+         })->then( sub {
+            matrix_send_room_text_message( $user, $room_id, body => "mice" )
+         })->then( sub {
+            do_request_json_for( $anonymous_user,
+               method => "GET",
+               uri    => "/api/v1/rooms/$room_id/messages",
+               params => {
+                  limit => "2",
+                  dir   => "b",
+               },
+            )
+         });
+      });
+   };
+
+sub register_anonymous_user
+{
+   my ( $http ) = @_;
+
+   $http->do_request_json(
+      method  => "POST",
+      uri     => "/v2_alpha/register",
+      content => {},
+      params  => {
+         kind => "guest",
+      },
+   )->then( sub {
+      my ( $body ) = @_;
+      my $access_token = $body->{access_token};
+
+      Future->done( User( $http, $body->{user_id}, $access_token, undef, undef, [], undef ) );
+   });
+}
+
+sub set_room_history_visibility
```

Comment:
(too late now as this is merged, but): This function could probably have been a `matrix_....` one that was exported so that later test files can use it too.